crates/noirc_evaluator/src/ssa_refactor/acir_gen/mod.rs

```diff
@@ -384,36 +423,150 @@ impl Context {
+                            dfg.type_of_value(dfg.instruction_results(instruction)[0]);
+                        let value = self.create_default_value(&result_type)?;
+                        self.define_result(dfg, instruction, value);
+                        return Ok(());
+                    }
+
+                    let value = match store_value {
+                        Some(store_value) => {
+                            let store_value = self.convert_value(store_value, dfg);
+                            AcirValue::Array(array.update(index, store_value))
+                        }
+                        None => array[index].clone(),
+                    };
+
+                    self.define_result(dfg, instruction, value);
+                    return Ok(());
+                }
+            }
+            AcirValue::DynamicArray(_) => (),
+        }
+
+        if let Some(store) = store_value {
+            self.array_set(instruction, array, index, store, dfg);
+        } else {
+            self.array_get(instruction, array, index, dfg);
         }
 
-        let value = match store_value {
-            Some(store_value) => {
-                let store_value = self.convert_value(store_value, dfg);
-                AcirValue::Array(array.update(index, store_value))
+        Ok(())
+    }
+
+    /// Generates a load opcode for the array
+    fn array_get(
+        &mut self,
+        instruction: InstructionId,
+        array: ValueId,
+        index: ValueId,
+        dfg: &DataFlowGraph,
+    ) {
+        let array = dfg.resolve(array);
+        let block_id = BlockId(array.to_usize() as u32);
+        if !self.initialized_arrays.contains(&block_id) {
+            match &dfg[array] {
+                Value::Array { array, .. } => {
+                    let values: Vec<AcirValue> =
+                        array.iter().map(|i| self.convert_value(*i, dfg)).collect();
+                    self.initialize_array(block_id, array.len(), &values);
+                }
+                _ => panic!("reading uninitialized array"),
             }
-            None => array[index].clone(),
+        }
+
+        let index_var = self.convert_value(index, dfg).into_var();
+        let read = self.acir_context.read_from_memory(block_id, &index_var);
+        let typ = match dfg.type_of_value(array) {
+            Type::Array(typ, _) => {
+                if typ.len() != 1 {
+                    unimplemented!(
+                        "Non-const array indices is not implemented for non-homogenous array"
+                    );
+                }
+                typ[0].clone()
+            }
+            _ => unreachable!("ICE - expected an array"),
         };
+        let typ = AcirType::from(typ);
+        self.define_result(dfg, instruction, AcirValue::Var(read, typ));
+    }
 
-        self.define_result(dfg, instruction, value);
-        Ok(())
+    // Copy the array and generates a write opcode on the new array
```

Comment:
Should add a note here stating that this is inefficient since it is initializing a new array and specify that this is not the way we want to do it in the future -- We should open up an issue for this too!

Reply:
#2014